src/infrastructure/acknowledgment.rs

```diff
@@ -0,0 +1,249 @@
+    }
+
+    #[test]
+    fn ack_bitfield_with_empty_receive() {
+        let handler = AcknowledgmentHandler::new();
+        assert_eq!(handler.ack_bitfield(), 0)
+    }
+
+    #[test]
+    fn ack_bitfield_with_some_values() {
+        let mut handler = AcknowledgmentHandler::new();
+        handler
+            .received_packets
+            .insert(0, ReceivedPacket::default());
+        handler
+            .received_packets
+            .insert(1, ReceivedPacket::default());
+        handler
+            .received_packets
+            .insert(3, ReceivedPacket::default());
+        assert_eq!(handler.ack_bitfield(), 0b1101)
+    }
+
+    #[test]
+    fn packet_is_not_acked() {
+        let mut handler = AcknowledgmentHandler::new();
+
+        handler.sequence_number = 0;
+        handler.process_outgoing(vec![1, 2, 3].as_slice(), OrderingGuarantee::None);
+        handler.sequence_number = 40;
+        handler.process_outgoing(vec![1, 2, 4].as_slice(), OrderingGuarantee::None);
+
+        static ARBITRARY: u16 = 23;
+        handler.process_incoming(ARBITRARY, 40, 0);
+
+        assert_eq!(
+            handler.dropped_packets(),
+            vec![SentPacket {
+                payload: vec![1, 2, 3].into_boxed_slice(),
+                ordering_guarantee: OrderingGuarantee::None,
+            }]
+        );
+    }
+
+    #[test]
+    fn acking_500_packets_without_packet_drop() {
+        let mut handler = AcknowledgmentHandler::new();
+        let mut other = AcknowledgmentHandler::new();
+
+        for i in 0..500 {
+            handler.sequence_number = i;
+            handler.process_outgoing(vec![1, 2, 3].as_slice(), OrderingGuarantee::None);
+
+            other.process_incoming(i, handler.remote_sequence_num(), handler.ack_bitfield());
+            handler.process_incoming(i, other.remote_sequence_num(), other.ack_bitfield());
+        }
+
+        assert_eq!(handler.dropped_packets().len(), 0);
+    }
+
+    #[test]
+    fn acking_many_packets_with_packet_drop() {
+        let mut handler = AcknowledgmentHandler::new();
+        let mut other = AcknowledgmentHandler::new();
+
+        let mut drop_count = 0;
+
+        for i in 0..100 {
+            handler.process_outgoing(vec![1, 2, 3].as_slice(), OrderingGuarantee::None);
+            handler.sequence_number = i;
+
+            // dropping every 4th with modulo's
+            if i % 4 == 0 {
+                debug!("Dropping packet: {}", drop_count);
+                drop_count += 1;
+            } else {
+                // We send them a packet
+                other.process_incoming(i, handler.remote_sequence_num(), handler.ack_bitfield());
+                // Skipped: other.process_outgoing
+                // And it makes it back
+                handler.process_incoming(i, other.remote_sequence_num(), other.ack_bitfield());
+            }
+        }
+
+        // TODO: Is this what we want?
+        //        assert_eq!(handler.dropped_packets.len(), 25);
```

Comment:
I think this is exactly what we want if we have 100 packets and we divide that by 4 we will get 25. So we should have 25 dropped packets over 100 packets. If you don't trust the modulo operation please try the following code out:

```rust

let mut counter = 1;
        for i in 0..100 {
            handler.process_outgoing(vec![1, 2, 3].as_slice(), OrderingGuarantee::None);
            handler.sequence_number = i;

            // dropping every 4th with modulo's
            if counter == 4 {
                println!("Dropping packet: {}", drop_count);
                drop_count += 1;

                counter = 1;
            } else {
                // We send them a packet
                other.process_incoming(i, handler.remote_sequence_num(), handler.ack_bitfield());
                // Skipped: other.process_outgoing
                // And it makes it back
                handler.process_incoming(i, other.remote_sequence_num(), other.ack_bitfield());

                counter += 1;
            }
        }
```
This will drop every fourth packet for sure.

Reply:
This goes back to the discussion we were having the other day on discord (plus I know how modulo works :P). The reason why we would only get 17 dropped packets in this case is that the algorithm waits until it sees 32 packets that haven't acked that sequence number. So at received packet 100, which is the last packet received from the remote host, we have an ack which looks like `01110111011101110111011101110111`. <- This is where those extra 8 packets are. We haven't yet determined that they have been dropped because they haven't "fallen off the end"